libvast/src/system/index.cpp

```diff
@@ -208,49 +208,149 @@ filesystem_actor& partition_factory::filesystem() {
-    return job.sender == sender;
-  });
-  return num;
+size_t pending_queue::num_queries() const {
+  return queries.size();
+}
+
+[[nodiscard]] uuid pending_queue::create_query_id() const {
+  auto query_id = uuid::random();
+  // Ensure the query id is unique.
+  while (queries.find(query_id) != queries.end() || query_id == uuid::nil())
+    query_id = uuid::random();
+  return query_id;
 }
 
-std::optional<query_backlog::job> query_backlog::take_next() {
-  if (!normal.empty()) {
-    auto result = normal.front();
-    normal.pop_front();
-    return result;
+[[nodiscard]] caf::error pending_queue::insert(query_state&& query_state,
+                                               std::vector<uuid>&& candidates) {
+  auto qid = query_state.query.id;
+  auto [it, emplace_success] = queries.emplace(qid, std::move(query_state));
+  if (!emplace_success)
+    return caf::make_error(ec::unspecified, "A query with this ID exists "
+                                            "already");
+  for (const auto& cand : candidates) {
+    auto it = std::find_if(partitions.begin(), partitions.end(), [&](auto& x) {
+      return x.partition == cand;
+    });
+    if (it != partitions.end()) {
+      it->queries.push_back(qid);
+      continue;
+    }
+    it = std::find_if(inactive_partitions.begin(), inactive_partitions.end(),
+                      [&](auto& x) {
+                        return x.partition == cand;
+                      });
+    if (it != inactive_partitions.end()) {
+      it->queries.push_back(qid);
+      partitions.push_back(std::move(*it));
+      inactive_partitions.erase(it);
+      continue;
+    }
+    partitions.push_back(pending_queue::pq{cand, std::vector{qid}});
   }
-  if (!low.empty()) {
-    auto result = low.front();
-    low.pop_front();
-    return result;
+  // TODO: Insertion sort should be better.
+  std::sort(partitions.begin(), partitions.end());
+  return caf::none;
+}
+
+[[nodiscard]] caf::error
+pending_queue::activate(const uuid& qid, uint32_t num_partitions) {
+  auto it = queries.find(qid);
+  if (it == queries.end())
+    return caf::make_error(ec::unspecified, "cannot activate unknown query");
+  it->second.requested_partitions += num_partitions;
+  // Go over all currently inactive partitions and splice those relevant for
+  // `qid` back into the active queue.
+  auto new_inactive = std::vector<pending_queue::pq>{};
+  std::partition_copy(std::make_move_iterator(inactive_partitions.begin()),
+                      std::make_move_iterator(inactive_partitions.end()),
+                      std::back_inserter(partitions),
+                      std::back_inserter(new_inactive), [&](const auto& p) {
+                        return std::find(p.queries.begin(), p.queries.end(),
+                                         qid)
+                               != p.queries.end();
+                      });
+  inactive_partitions = std::move(new_inactive);
+  std::sort(partitions.begin(), partitions.end());
```

Comment:
Unfortunately not, because we change priority values for existing items when an exporter requests another batch. Supporting that with a heap comes down to the same work but is more cumbersome to implement.